src/Compilers/Core/MSBuildTask/LocateEditorConfigFiles.cs

```diff
@@ -0,0 +1,122 @@
+            {
+                // If we're being passed files that don't exist, we'll just skip them and let the compiler handle the actual reporting.
+                if (directory.Exists)
+                {
+                    if (directoriesAddedToQueue.Add(directory.FullName))
+                    {
+                        directoriesToVisit.Enqueue(directory);
+                    }
+                }
+            }
+
+            foreach (var inputFile in InputFiles)
+            {
+                addNewDirectoryIfNotAlreadyAdded(new DirectoryInfo(Path.GetDirectoryName(inputFile.ItemSpec)));
+            }
+
+            while (directoriesToVisit.Count > 0)
+            {
+                var directory = directoriesToVisit.Dequeue();
+                var editorConfigFilePath = Path.Combine(directory.FullName, ".editorconfig");
+                bool isRootEditorConfig = false;
+
+                if (File.Exists(editorConfigFilePath))
+                {
+                    editorConfigFiles.Add(new TaskItem(editorConfigFilePath));
+
+                    // We will attempt to determine if this is a root .editorconfig, so we can stop enumerating farther if necessary.
+                    // We will do no further interpretation of the .editorconfig.
+
+                    isRootEditorConfig = FileIsRootEditorConfig(editorConfigFilePath);
+                }
+
+                if (!isRootEditorConfig && directory.Parent != null)
+                {
+                    addNewDirectoryIfNotAlreadyAdded(directory.Parent);
+                }
+            }
+
+            EditorConfigFiles = editorConfigFiles.ToArray();
+        }
+
+        // These Regexes are the regexes from Microsoft.VisualStudio.CodingConventions.EditorConfig.
+        private static Regex _sectionMatcher = new Regex(@"^\s*\[(([^#;]|\\#|\\;)+)\]\s*([#;].*)?$", RegexOptions.Compiled);
+        private static Regex _propertyMatcher = new Regex(@"^\s*([\w\.\-_]+)\s*[=:]\s*(.*?)\s*([#;].*)?$", RegexOptions.Compiled);
+
+        internal static bool FileIsRootEditorConfig(string editorConfigFilePath)
+        {
+            using (FileStream fileStream = new FileStream(editorConfigFilePath, FileMode.Open, FileAccess.Read, FileShare.Read))
+            using (StreamReader reader = new StreamReader(fileStream))
+            {
+                string line;
+                while ((line = reader.ReadLine()) != null)
+                {
+                    // Once we're in a section, we don't need to go looking for root = true anymore
+                    if (_sectionMatcher.IsMatch(line))
+                    {
+                        return false;
+                    }
+
+                    var propMatches = _propertyMatcher.Matches(line);
+                    if (propMatches.Count > 0 && propMatches[0].Groups.Count > 1)
+                    {
+                        var key = propMatches[0].Groups[1].Value;
+                        var value = propMatches[0].Groups[2].Value;
```

Comment:
The guard in the `if` block checks for `Count > 1` but this unconditionally references `Groups[2]`.  One of these is incorrect.